apps_download/models/product_product.py

```diff
@@ -0,0 +1,107 @@
+                for child in children:
+                    if not child.dependent_product_ids:
+                        continue
+                    if child in product_dependent_ids:
+                        raise ValidationError(
+                            _('Error: You cannot create recursive dependency.')
+                        )
+                    product_dependent_ids += child
+                    child_dependency_check(product_dependent_ids,
+                                           child.dependent_product_ids)
+                return res
+
+            child_dependency_check(product, product.dependent_product_ids)
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+
+        def child_dependency(children):
+            res = self.env['product.product']
+            for child in children:
+                if not child.dependent_product_ids:
+                    continue
+                res += child.dependent_product_ids
+                res += child_dependency(child.dependent_product_ids)
+            return res
+        for product in self:
+            ret_val[product.id] = product.dependent_product_ids
+            if product.dependent_product_ids:
+                ret_val[product.id] += child_dependency(
+                    product.dependent_product_ids)
+        return ret_val
+
+    @api.multi
+    def generate_zip_file(self):
+        for product in self:
+            if not product.module_path:
+                continue
+            tmpdir = tempfile.mkdtemp()
+            tmpdir2 = tempfile.mkdtemp()
+            dependent_product_ids = product.create_dependency_list(
+                )[product.id]
+            for dependent_product in dependent_product_ids:
+                if not dependent_product.module_path:
+                    continue
+                subprocess.Popen(['cp', '-r', dependent_product.module_path,
+                                  tmpdir], stdout=subprocess.PIPE)
+            subprocess.Popen(['cp', '-r', product.module_path, tmpdir],
+                             stdout=subprocess.PIPE)
+            time_value = time.strftime(
+                '_%y%m%d_%H%M%S')
+
+            tmpzipfile = os.path.join(tmpdir2, product.name) + time_value
+            shutil.make_archive(tmpzipfile, 'zip', tmpdir)
+            tmpzipfile = tmpzipfile + '.zip'
+            with open(tmpzipfile, "rb") as fileobj:
+                try:
+                    data_encode = base64.encodestring(fileobj.read())
+                    self.env['ir.attachment'].create({
+                        'datas': data_encode,
+                        'datas_fname': tmpzipfile,
+                        'type': 'binary',
+                        'name': product.name + time_value + '.zip',
+                        'res_model': product._name,
+                        'res_id': product.id,
+                        'product_downloadable': True,
+                    })
+                except:
```

Comment:
You should only except for the specific things you expect. In this case, I'm assuming `OSError`

Reply:
Thinking further - are we sure we want to cover exceptions here? Why is this considered normal enough to silence?

Silencing errors means that they will not bubble into error traps, such as Sentry. This is a primary mechanism of identifying errors in production systems - logs are meant for once you identify the errors. Moving an error into a log means that the log will never be read for the error, because the administrator will not know it exists.

Reply:
@lasley I think the exception is because this method is going to be called mostly from the Cron job and that is why there is exception handling so is one product will have some error then it will not stop creating files for other products.

@agenterp  Correct me if I am wrong.

Reply:
Hmmm but why would the execution fail? IMO If the product fails once, it will fail every time & should be mitigated. 

Can you add something to the roadmap in ReadMe for the creation of an error queue to handle these? IMO if we are ignoring the failure because we expect them, we need an error trap and process for mitigation of the error. I feel adding it to the scope of this PR is a bit much though, so the readme will do I think

Reply:
Agree: at least a note in the roadmap

Reply:
I concur with you idea to add into the roadmap and not adding the functionality in the same PR.
I have just done the commit for the roadmap.